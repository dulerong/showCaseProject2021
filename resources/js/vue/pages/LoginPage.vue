<template lang="pug">
v-card(class="mx-auto pa-4" style="width: 375px")
  h1 {{ greetingMessage }},
  h1 Login Page
  v-form
    v-text-field(
      label="User Name"
      v-model="username"
      hint="Showcase use only (You don't have to enter username)"
      clearable
    )
    v-text-field(
      label="Password"
      v-model="password"
      :type="showPassword? 'text' : 'password' "
      :append-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
      hint="Showcase use only (You don't have to enter password)"
      @click:append="showPassword = !showPassword"
      clearable
    )
  v-row(no-gutters justify="space-between" class="mt-1")
    v-btn(depressed :loading="loading" @click="login") 
      v-icon(left) mdi-login
      | Login
    v-btn(outlined)
      v-icon(left) mdi-clipboard-account
      | Sign up
</template>

<script>
export default {
  created() {
    this.greetingMessage = this.makeGreetMessage(this.$_hour_right_now)
  },
  data: () => ({
    greetingMessage: null,
    username: null,
    password: null,
    showPassword: false,
    loading: false,
  }),
  methods: {
    makeGreetMessage(time) {
      switch(true) {
        case time < 12:
          return 'Good morning'
        case time >= 12 && time < 18:
          return 'Good afternoon'
        case time >= 18 && time <= 23:
          return 'Good evening'
      }
    },
    login() {
      this.loading = true
      setTimeout(() => {
        this.loading = false
        this.$router.push({ name: this.$_routePath.home.name })
      }, 2500)
    }
  }
}
</script>