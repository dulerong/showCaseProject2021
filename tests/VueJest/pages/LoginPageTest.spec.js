import { mount, createLocalVue } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex from 'vuex'

import LoginPage from '~pages/LoginPage'

describe('Login Page', () => {
  const localVue = createLocalVue()
  const mockRoutes = {
    home: {
      name: 'home'
    }
  }

  localVue.use(Vuex)

  let vuetify
  let store
  let notification
  let notificationMutations

  beforeEach(() => {
    vuetify = new Vuetify()
    notificationMutations = {
      showNotification: jest.fn(),
    }
    notification = {
      namespaced: true,
      mutations: notificationMutations
    }
    store = new Vuex.Store({
      modules: {
        notification
      }
    })
  })

  afterEach(() => {    
    jest.clearAllMocks()
  })

  const mountFunction = options => {
    return mount(LoginPage, {
      localVue,
      vuetify,
      store,
      mocks: {
        $t: msg => msg,
        $router: [],
        $_routePath: mockRoutes
      },
      ...options
    })
  }

  it('Page renders properly', () => {
    const wrapper = mountFunction()

    expect(wrapper.exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Computed greetingMessage returns correct message', () => {
    const localThis = (mockHour) => ({
      $t: msg => msg,
      $_hour_right_now: mockHour
    })

    expect(LoginPage.computed.greetingMessage.call(localThis(5)))
      .toEqual('login.greeting.morning')

    expect(LoginPage.computed.greetingMessage.call(localThis(14)))
      .toEqual('login.greeting.afternoon')

    expect(LoginPage.computed.greetingMessage.call(localThis(20)))
      .toEqual('login.greeting.evening')
  })

  it('Method login works correctly', async () => {
    jest.useFakeTimers();
    const wrapper = mountFunction()

    await wrapper.vm.login()
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 2500)
    expect(wrapper.vm.loading).toBeTruthy()

    jest.runAllTimers()
    expect(wrapper.vm.loading).toBeFalsy()
    expect(wrapper.vm.$router).toStrictEqual([
      { name: wrapper.vm.$_routePath.home.name }
    ])
    expect(notificationMutations.showNotification)
      .toHaveBeenCalledWith({}, { message: 'login.greeting.welcome' })
  })
})