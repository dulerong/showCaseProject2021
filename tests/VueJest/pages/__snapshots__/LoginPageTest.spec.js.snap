// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Tests Page renders properly 1`] = `
<div class="container">
  <div class="stripe"></div><img src="/images/logo/ikeda.svg">
  <h1 class="my-5 blue-grey--text">Login Page</h1>
  <div class="pa-5 v-card v-sheet theme--light" style="width: 440px;">
    <form novalidate="novalidate" class="v-form">
      <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
        <div class="v-input__control">
          <div class="v-input__slot field">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span>​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-8" class="v-label theme--light" style="left: 0px; position: absolute;">ログインID</label><input id="input-8" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
        <div class="v-input__control">
          <div class="v-input__slot field">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span>​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-11" class="v-label theme--light" style="left: 0px; position: absolute;">パスワード</label><input id="input-11" type="password"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input my-0 theme--light v-input--selection-controls v-input--checkbox">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"></i><input aria-checked="false" id="input-14" role="checkbox" type="checkbox" value="">
              <div class="v-input--selection-controls__ripple"></div>
            </div><label for="input-14" class="v-label theme--light" style="left: 0px; position: relative;">ログインしたままにする</label>
          </div>
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div><button type="submit" disabled="disabled" class="text-h4 v-btn v-btn--block v-btn--disabled v-btn--has-bg v-btn--tile theme--light v-size--default" style="height: 50px;"><span class="v-btn__content">LOGIN</span></button>
    </form>
  </div>
  <h4 class="my-5 text-decoration-underline primary--text">＜パスワードを忘れた方はこちらへ</h4>
</div>
`;
